drivers/usbc/tcpc/ucpd_stm32.c

```diff
@@ -0,0 +1,1463 @@
+					ucpd_set_tx_state(info->dev,
+							  STATE_IDLE);
+				} else {
+					/*
+					 * Tx attempt failed. Remain in this
+					 * state, but trigger new tx attempt.
+					 */
+					msg_src = TX_MSG_TCPM;
+					data->tx_retry_count++;
+				}
+			} else {
+				enum tcpc_alert status;
+
+				status = (info->evt & UCPD_EVT_TX_MSG_FAIL) ?
+					 TCPC_ALERT_TRANSMIT_MSG_FAILED :
+					 TCPC_ALERT_TRANSMIT_MSG_DISCARDED;
+				ucpd_set_tx_state(info->dev, STATE_IDLE);
+				info->handler(info->dev, info->data, status);
+			}
+		}
+		break;
+
+	case STATE_ACTIVE_CRC:
+		if (info->evt & (UCPD_EVT_TX_MSG_SUCCESS |
+				 UCPD_EVT_TX_MSG_FAIL |
+				 UCPD_EVT_TX_MSG_DISC)) {
+			ucpd_set_tx_state(info->dev, STATE_IDLE);
+			if (info->evt & UCPD_EVT_TX_MSG_FAIL) {
+				LOG_INF("ucpd: Failed to send GoodCRC!");
+			} else if (info->evt & UCPD_EVT_TX_MSG_DISC) {
+				LOG_INF("ucpd: GoodCRC message discarded!");
+			}
+		}
+		break;
+
+	case STATE_WAIT_CRC_ACK:
+		if (info->evt & UCPD_EVT_RX_GOOD_CRC &&
+		    data->ucpd_crc_id == data->msg_id_match) {
+			/* GoodCRC with matching ID was received */
+			info->handler(info->dev, info->data,
+				      TCPC_ALERT_TRANSMIT_MSG_SUCCESS);
+			ucpd_set_tx_state(info->dev, STATE_IDLE);
+		} else if (info->evt & UCPD_EVT_RX_GOOD_CRC) {
+			/* GoodCRC w/out match or timeout waiting */
+			if (data->tx_retry_count < data->tx_retry_max) {
+				ucpd_set_tx_state(info->dev, STATE_ACTIVE_TCPM);
+				msg_src = TX_MSG_TCPM;
+				data->tx_retry_count++;
+			} else {
+				ucpd_set_tx_state(info->dev, STATE_IDLE);
+				info->handler(info->dev, info->data,
+					      TCPC_ALERT_TRANSMIT_MSG_FAILED);
+			}
+		} else if (info->evt & UCPD_EVT_RX_MSG) {
+			/*
+			 * In the case of a collsion, it's possible the port
+			 * partner may not send a GoodCRC and instead send the
+			 * message that was colliding. If a message is received
+			 * in this state, then treat it as a discard from an
+			 * incoming message.
+			 */
+			info->handler(info->dev, info->data,
+				      TCPC_ALERT_TRANSMIT_MSG_DISCARDED);
+			ucpd_set_tx_state(info->dev, STATE_IDLE);
+		}
+		break;
+
+	case STATE_HARD_RESET:
+		if (info->evt & UCPD_EVT_HR_DONE) {
+			/* HR complete, reset tx state values */
+			ucpd_set_tx_state(info->dev, STATE_IDLE);
+			data->ucpd_tx_request = 0;
+			data->tx_retry_count = 0;
+		} else if (info->evt & UCPD_EVT_HR_FAIL) {
+			ucpd_set_tx_state(info->dev, STATE_IDLE);
+			data->ucpd_tx_request = 0;
+			data->tx_retry_count = 0;
+		}
+		break;
+	}
+
+	/* If msg_src is valid, then start transmit */
+	if (msg_src > TX_MSG_NONE) {
+		ucpd_start_transmit(info->dev, msg_src);
+	}
+}
+
+/**
+ * @brief Alert handler
+ */
+static void ucpd_alert_handler(struct k_work *item)
+{
+	struct alert_info *info = CONTAINER_OF(item, struct alert_info, work);
+	struct tcpc_data *data = info->dev->data;
+
+	if (info->evt & UCPD_EVT_EVENT_CC) {
+		info->handler(info->dev, info->data, TCPC_ALERT_CC_STATUS);
```

Comment:
it is good practice to check for valid handler against NULL before calling.
as there are multiple places where handler is used, may adding new static function may help.

static void ucpd_nofity_handler(dev, data, status)
{
        if (dev->handler) {
                   dev->handler(dev, data, status);
        }
}

Reply:
Done